.github/workflows/build_charmcraft_cache.yaml

```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+name: Build Charmcraft Cache
+
+on:
+  workflow_call:
+    inputs:
+      runs-on:
+        type: string
+        description: Image runner for building the images
+        default: ubuntu-22.04
+
+jobs:
+  build-charmcraft-cache:
+    name: Build Charmcraft Cache
+    runs-on: ${{ inputs.runs-on }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Setup operator environment
+        uses: charmed-kubernetes/actions-operator@main
+        with:
+          channel: latest/stable
+      - name: Generate charmcraft cache key
+        run: |
+          CHARM_NAME=$(yq .name ./metadata.yaml)
+          CHARMCRAFT_BASE="$(yq -o=j -I=0 '.bases | sort_keys(..)' ./charmcraft.yaml)"
+          CHARMCRAFT_BASE=$(printf %s "$CHARMCRAFT_BASE"|jq -sRr @uri)
+          CHARMCRAFT_CACHE_KEY_BASE="charmcraft-cache?name=$CHARM_NAME&bases=$CHARMCRAFT_BASE"
+          echo "CHARMCRAFT_CACHE_KEY=$CHARMCRAFT_CACHE_KEY_BASE&date=$(date +%Y-%m-%d)" >> $GITHUB_ENV
+          echo 'CHARMCRAFT_CACHE_ALT_KEYS<<EOF' >> $GITHUB_ENV 
+          for d in {1..2}
+            do echo "$CHARMCRAFT_CACHE_KEY_BASE&date=$(date -d"-$d days" +%Y-%m-%d)" >> $GITHUB_ENV
+          done
+          echo 'EOF' >> $GITHUB_ENV
+      - name: Build Charm
+        run: |
+          charmcraft pack
```

Comment:
If we build the charm here, to avoid packing it again during integration testing, I suggest we upload the artifact here and download it later, packing once to avoid dependencies changing between builds

Reply:
The problem is that, one charmcraft takes about 1 GB space, and we only has 10 GB per repository. So we can not save a cache for each pull request, all pull request has to share the cache build on the main branch.

Reply:
The issue with that is that there's no guarantee that the image in CI is the same as the image built during publish

Reply:
I think I didn't explain myself very well. I meant the .charm file resulting from the pack that we are already uploading as an artifact during integration testing

Reply:
Okay, I will come up with a solution that reuse the charm from the CI workflows in the publish workflow.

Reply:
Updated, the charm and images are now reused in the publish workflow.

Reply:
We are also packing in the integration_test.yaml. Should the `charmcraft pack` be removed from there?